src/tactic/spass/examples.lean

```diff
@@ -0,0 +1,115 @@
+  example : (∀ x, p x) → (∀ x, p x → q x) → ∀ x, q x := by spass
+  example : (∃ x, p x) → (∀ x, p x → q x) → ∃ x, q x := by spass
+  example : (¬ ∀ x, ¬ p x) → (∀ x, p x → q x) → (∀ x, ¬ q x) → false := by spass
+  example : (p a) → (p a → false) → false := by spass
+  example : (¬ (P → Q ∨ R)) → (¬ (Q ∨ ¬ R)) → false := by spass
+  example : (P → Q ∨ R) → (¬ Q) → P → R := by spass
+  example : (∃ x, p x → P) ↔ (∀ x, p x) → P := by spass
+  example : (∃ x, P → p x) ↔ (P → ∃ x, p x) := by spass
+
+end
+
+section
+
+  /- Examples from the Handbook of Practical Logic and Automated Reasoning. -/
+
+  variables (α : Type) [inhabited α]
+
+  lemma gilmore_1 {F G H : α → Prop} :
+    ∃ x, ∀ y z,
+        ((F y → G y) ↔ F x) ∧
+        ((F y → H y) ↔ G x) ∧
+        (((F y → G y) → H y) ↔ H x)
+        → F z ∧ G z ∧ H z := by spass
+
+  lemma gilmore_6 {F G : α → α → Prop} {H : α → α → α → Prop} :
+  ∀ x, ∃ y,
+    (∃ u, ∀ v, F u x → G v u ∧ G u x)
+     → (∃ u, ∀ v, F u y → G v u ∧ G u y) ∨
+         (∀ u v, ∃ w, G v u ∨ H w y u → G u w) := by spass
+
+  lemma gilmore_8 {G : α → Prop} {F : α → α → Prop} {H : α → α → α → Prop} :
+    ∃ x, ∀ y z,
+      ((F y z → (G y → (∀ u, ∃ v, H u v x))) → F x x) ∧
+      ((F z x → G x) → (∀ u, ∃ v, H u v z)) ∧
+      F x y → F z z := by spass
+
+  lemma Manthe_and_Bry (agatha butler charles : α)
+  (lives : α → Prop) (killed hates richer : α → α → Prop) :
+     lives agatha ∧ lives butler ∧ lives charles ∧
+     (killed agatha agatha ∨ killed butler agatha ∨
+      killed charles agatha) ∧
+     (∀ x y, killed x y → hates x y ∧ ¬ richer x y) ∧
+     (∀ x, hates agatha x → ¬ hates charles x) ∧
+     (hates agatha agatha ∧ hates agatha charles) ∧
+     (∀ x, lives(x) ∧ ¬ richer x agatha → hates butler x) ∧
+     (∀ x, hates agatha x → hates butler x) ∧
+     (∀ x, ¬ hates x agatha ∨ ¬ hates x butler ∨ ¬ hates x charles)
+     → killed agatha agatha ∧
+         ¬ killed butler agatha ∧
+         ¬ killed charles agatha := by spass
+
+  lemma Knights_and_Knaves (me : α) (knight knave rich poor : α → α)
+    (a_truth says : α → α → Prop) (and : α → α → α) :
+    ( (∀ X Y, ¬ a_truth (knight X) Y ∨ ¬ a_truth (knave X) Y ) ∧
+      (∀ X Y, a_truth (knight X) Y ∨ a_truth (knave X) Y ) ∧
+      (∀ X Y, ¬ a_truth (rich X) Y ∨ ¬ a_truth (poor X) Y ) ∧
+      (∀ X Y, a_truth (rich X) Y ∨ a_truth (poor X) Y ) ∧
+      (∀ X Y Z, ¬ a_truth (knight X) Z ∨ ¬ says X Y ∨ a_truth Y Z ) ∧
+      (∀ X Y Z, ¬ a_truth (knight X) Z ∨ says X Y ∨ ¬ a_truth Y Z ) ∧
+      (∀ X Y Z, ¬ a_truth (knave X) Z ∨ ¬ says X Y ∨ ¬ a_truth Y Z ) ∧
+      (∀ X Y Z, ¬ a_truth (knave X) Z ∨ says X Y ∨ a_truth Y Z ) ∧
+      (∀ X Y Z, ¬ a_truth (and X Y) Z ∨ a_truth X Z ) ∧
+      (∀ X Y Z, ¬ a_truth (and X Y) Z ∨ a_truth Y Z ) ∧
+      (∀ X Y Z, a_truth (and X Y) Z ∨ ¬ a_truth X Z ∨ ¬ a_truth Y Z ) ∧
+      (∀ X, ¬ says me X ∨ ¬ a_truth (and (knave me) (rich me)) X ) ∧
+      (∀ X, says me X ∨ a_truth (and (knave me) (rich me)) X ) ) → false := by spass
```

Comment:
I get a deterministic timeout error here.

Reply:
I forgot to note that this one was tested under extended time settings. It takes about 15s on my i7 laptop with time limit = 200000. I'll leave a note and comment it out.

Reply:
Out of curiosity, what's the bottleneck, SPASS or the proof reconstruction?

Reply:
The latter, by far. Even in the examples where the tactic struggles, SPASS is nearly instantaneous.

Reply:
And 99% of that time is eaten up by `to_expr`. Not surprising given the amount of computation by reflection that's taking place, but getting rid of it might require a complete redesign.

Reply:
`to_expr`? That's the elaborator. Maybe it's possible to skip the elaborator entirely and send the expression straight to the kernel? I think some variant on `exact` can achieve this, or `add_decl` in the worst case (which is possible even in a `by spass` style tactic, if you create an auxiliary definition, send it straight to the kernel, then stick the resulting constant in for the goal).

Reply:
@digama0 The final expr which gets applied to the original goal includes a term of the form `r.run m`, where `r` is the recipe that includes instructions on how to manipulate clauses, `m` is the initial set of CNF clauses, and the success of the proof depends on `r.run m` reducing to the empty clause `some []`. When I tried to avoid `to_expr` by using `return '(...)` it didn't work, and I thought the elaborator is necessary for getting the kernel computation to work. But maybe this isn't the case?